Wikipedia/Code/WMFAppViewController+Extensions.swift

```diff
@@ -283,29 +283,45 @@ extension WMFAppViewController: WKWatchlistDelegate {
     }
 
     public func watchlistUserDidTapAddLanguage(from viewController: UIViewController, viewModel: WKWatchlistFilterViewModel) {
-        displayAddLanguageSettingsForWatchlistFilters(from: viewController) {
-            // TODO
-            // Handle reloading filter view model projects when changing languages
-            viewModel.reload()
+        let languagesController = WMFLanguagesViewController(nibName: "WMFLanguagesViewController", bundle: nil)
+        languagesController.title = CommonStrings.wikipediaLanguages
+        languagesController.apply(theme)
+        languagesController.delegate = self
+        languagesController.showAllLanguages = true
+        languagesController.showPreferredLanguages = false
+        languagesController.showNonPreferredLanguages = false
+
+        languagesController.userLanguageSelectionBlock = { [weak self, weak viewModel] in
+            guard let self = self else { return }
+
+            // From `ViewControllerRouter`
+            let dataStore = self.dataStore
+            let appLanguages = dataStore.languageLinkController.preferredLanguages
+            var localizedProjectNames = appLanguages.reduce(into: [WKProject: String]()) { result, language in
+                guard let wikimediaProject = WikimediaProject(siteURL: language.siteURL, languageLinkController: dataStore.languageLinkController), let wkProject = wikimediaProject.wkProject else {
+                    return
+                }
+
+                result[wkProject] = wikimediaProject.projectName(shouldReturnCodedFormat: false)
+            }
+            localizedProjectNames[.wikidata] = WikimediaProject.wikidata.projectName(shouldReturnCodedFormat: false)
+            localizedProjectNames[.commons] = WikimediaProject.commons.projectName(shouldReturnCodedFormat: false)
```

Comment:
An alternative you could do here is to change [this line](https://github.com/wikimedia/wikipedia-ios/blob/main/Wikipedia/Code/ViewControllerRouter.swift#L237) to `dataStore.languageLinkController.allLanguages`. The filter localized strings will be populated with every possible language name upon init, and then you won't have to bother with updating it at this point after they add a language.

I understand the preference to keep localized strings limited to only will display on screen though, so not a must.

Reply:
No strong preference for me, just thought I'd mention it in case you were drawn more to it. I'm going to go ahead and approve since it's working well and there's good reason for the current approach.